File: catkin_ws/src/cone_detection_pipeline/src/image_saver.cpp
#include "ros/ros.h"
#include <ros/package.h>
#include <std_msgs/Bool.h>
#include <image_transport/image_transport.h>
#include <cv_bridge/cv_bridge.h>
#include <sensor_msgs/image_encodings.h>
#include <opencv2/imgproc/imgproc.hpp>


#include "topic_names.h"

// Declare "member" variables
image_transport::Subscriber left_sub;
image_transport::Subscriber right_sub;

// Converting ROS images into OpenCV images, using cv_bridge
namespace cv_bridgeI {

class CvImage
{
public:
  std_msgs::Header header;
  std::string encoding;
  cv::Mat image;
};

typedef boost::shared_ptr<CvImage> CvImagePtr;
typedef boost::shared_ptr<CvImage const> CvImageConstPtr;

}

#define SKIP 30
#define SAVE 60
// CV image pointers
cv_bridgeI::CvImagePtr LeftPtr;
cv_bridgeI::CVImagePtr RightPtr;
int savedCountL = 0;
int savedCountR = 0;
int skipCountL = SKIP;
int skipCountR = SKIP;

void leftImageCallback(const sensor_msgs::ImageConstPtr& msg){
	if skipCountL <= 0 {
	//Convert to opencv image using cv bridge
		LeftPtr = cv_bridge::toCvCopy(msg);
		
	}
}
void rightImageCallback(const sensor_msgs::ImageConstPtr& msg){
	//Convert to opencv image using cv bridge
	RightPtr = cv_bridge::toCvCopy(msg);
	//Set flag to true
	new_right = true;
}

void run_pipeline(const ros::TimerEvent& timer){
	if (new_left && new_right) {
		//GIVE IMAGES TO CLASS
		new_left = false;
		new_right = false;		
		//RUN THE PIPELINE
			
	}	
}

int main(int argc, char* argv[]) {
	// Initialise the node
	ros::init(argc, argv, "mur_stereo_ros_node");
	ros::NodeHandle nh;
    
	image_transport::ImageTransport it(nh);
	left_sub = it.subscribe("/CameraLeft/image_raw", 1, leftImageCallback);
	right_sub = it.subscribe("/CameraRight/image_raw", 1, rightImageCallback);
	ros::Timer timer = nh.createTimer(ros::Duration(0.05), run_pipeline);
    
	// Initialise a publisher
	cones_pub = nh.advertise<cone_msg>(CONE_DETECTED_TOPIC, 10, false);

	// Spin as a single-threaded node
	ros::spin();

	return 0;
}

File: catkin_ws/src/cone_detection_pipeline/src/ImageSaver.py
#! /usr/bin/env python

import rospy

from sensor_msgs.msg import Image

import numpy as np
import cv2 as cv
from cv_bridge import CvBridge


# Please add a path!
SAVE_IMAGE_PATH = ""

SAVE = 30
SKIP = 30

class ImageSaver:
    def __init__(self):
        self.r_sub = rospy.Subscriber("/CameraRight/image_raw", Image, self.rightInput)
        self.l_sub = rospy.Subscriber("/CameraLeft/image_raw", Image, self.leftInput)

        self.r_saved = 0;
        self.l_saved = 0;
        self.r_to_skip = SKIP;
        self.l_to_skip = SKIP;

        self.cv_bridge = CvBridge()

    def rightInput(self, msg):
        if self.r_saved < SAVE and self.r_to_skip <= 0:
            cv.imwrite(SAVE_IMAGE_PATH + "/right/" + str(self.r_saved) + ".png", self.cv_bridge.imgmsg_to_cv2(msg))
            self.r_saved += 1
            self.r_to_skip = SKIP
        else:
            self.r_to_skip -= 1
    
    def leftInput(self, msg):
        if self.l_saved < SAVE and self.l_to_skip <= 0:
            cv.imwrite(SAVE_IMAGE_PATH + "/left/" + str(self.l_saved) + ".png", self.cv_bridge.imgmsg_to_cv2(msg))
            self.l_saved += 1
            self.l_to_skip = SKIP
        else:
            self.l_to_skip -= 1
    

if __name__ == '__main__':
    # Initialise the node
    global node_name
    node_name = "ImageSaver"
    rospy.init_node(node_name)
    image_saver = ImageSaver()
    # Spin as a single-threaded node
    rospy.spin()

    # Close any OpenCV windows
    cv.destroyAllWindows()
